---
layout: tutorial_slides
logo: "GTN"

title: "Quality Control"
zenodo_link: "https://doi.org/10.5281/zenodo.61771"
questions:
  - Why quality control?
  - Do we need biological knowledge?
  - How to control quality of NGS data?
  - What are the quality parameters to check for each dataset?
  - How to improve the quality of a sequence dataset?
objectives:
  - Manipulate FASTQ files
  - Control quality from a FASTQ file
  - Use FastQC tool
  - Understand FastQC output
  - Use tools for quality correction
requirements:
time_estimation: "1h"
key_points:
  - Run quality control on every sequencing dataset before any other analyses
  - Choose QC parameters carefully
  - Re-run FastQC to check the impact of the quality control
contributors:
  - bebatut
  - gallardoalba
---

# Why Quality Control?

![Bungee jumping](../../images/quality-control/bungee.jpg)

---

### Do I need any knowledge about data generation protocols and biological background?

![Cartoon of different types of sequencing and where they appear in the genome. Bisulfite and ChIP-Seq have arrows pointing to nucleosomes. DNaseq-seq points to the region between nucleosomes. Hi-C and ChIA-PET point to the long range chromatin interactions. RNA-Seq points to a subset of the genome showing a promoter and transcribed region.](../../images/ecker_2012.png)

<small>[*Ecker et al, Nature, 2012*](https://www.ncbi.nlm.nih.gov/pubmed/22955614)</small>

???

- Many sequencing techniques
- [Huge diversity in protocols](https://www.illumina.com/science/sequencing-method-explorer.html)
- Knowledge about source of data (+expectations of how it looks) is important in QC process
- What is/isn't normal for your data

Segue: Might be concerned about different processing for each


---

## Results of *Homo sapiens* DNA sequencing experiments

![Human genome](../../images/quality-control/human_genome.png)

---

![Human genome](../../images/quality-control/retrovirus.jpg)

About 8% of our genome is composed of **sequences with viral origin**, relics of ancient infections that affected the primates' germ line along the last 100 million of years.

---

## Biological knowledge is **essential** for performing quality control

---

## Technical considerations

---

### Sequences: FASTA

```
>Identifier1 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
>Identifier2 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XX
```

???

- familiar if you've seen genomic data before
- Starts with greater-than sign
- Then an identifier with no spaces
- Everything after a space is a comment
- Newline
- Everything up to next greater-than is sequence data (wrapping unnecessary)

Segue: But this is just sequence, and we have data from a sequencer, which includes quality

---

### Sequences: FASTA

![FASTA sequence](../../images/quality-control/FASTA_example.png)

---

### Sequences: FASTQ - Quality score

The most extended format is Phred+33, which uses a score from 0 to 40 using ASCII 33 to 73.

![Quality scores](../../images/quality-control/quality_scores.png)

???

- Logarithmic scale


## Identifying Potential Quality Issues: FastQC

---

### Sequences: FASTQ

```
@Identifier1 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
@Identifier2 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
```

???

- Four lines
- @ + identifier on first line, just like fasta
- sequence
- +
- quality score characters

Segue: so what do the quality chars mean?

---

### Sequences: FASTQ

![FASTQ sequence](../../images/quality-control/FASTQ_example.png)

---


### Per-base Quality

![Fastqc quality score plot, most results are in the green region up until 30. The whiskers extend to the yellow region from the start, and after base 30 get progressively worse, goign to the worst possible score by the end. The boxes cover the yellow region by base 40.](../../images/per_base_sequence_quality_bad.png)

<i class="fa fa-thumbs-down"></i> Bad quality score

---

### Per-base Quality

![A graph in between the previous two, mostly green, but gts to yellow  by base 30 and red by base 33.](../../images/per_base_sequence_quality_medium.png)

<i class="fa fa-thumbs-up"></i> <i class="fa fa-thumbs-down"></i> Intermediate quality score

---

### Per-sequence Quality

![A per-sequence quality showing quality score distribution histogram with average quality per read plotted. Most reads pile up around quality 30, but another peak appears at 17. very few reads have quality less than 10.](../../images/per_sequence_quality.png)

---

### Per-base Sequence Content

![A line chart mapping sequence content across all bases with % of the four nucleotides plotted. The graph starts off very jagged and stabilises by base 12.](../../images/per_base_sequence_content.png)

???
The per-base sequence content highlights the proportion of each base in each position of a sequence for which each of the four DNA bases have been called. In a random library there would be little to no difference between the different bases of a sequence run. The relative amount of each base should reflect the overall amount of these bases, but in any case they should not be hugely imbalanced from one another. It is worth noting that some types of libraries will always produce biased sequence composition, normally at the start of the read. Libraries produced by priming with random hexamers (including nearly all RNA-Seq libraries) and those which were fragmented using transposases, inherit an intrinsic bias in the positions at which the reads start. This bias does not concern an absolute sequence, but instead provides an enrichment of a number of different K-mers at the 5’ end of the reads. Whilst this is a true technical bias, it isn’t something which can be corrected by trimming and in most cases doesn’t seem to adversely affect the downstream analysis. It will however produce a warning or error in this module.

There are a number of common scenarios for these issues:

- Over-represented sequences
- Biased fragmentation
- Biased composition libraries
- Aggressive adapter trimming

---

### Per-sequence GC content

![A line chart showing mean GC content and threoretical distribution as largely overlapping peaks.](../../images/per_sequence_GC_content.png)

???
The GC content distribution of most samples should follow a normal distribution. In some cases, a bi-modal distribution can be observed, especially for metagenomic data sets. An unusually shaped distribution could indicate a contaminated library or some other kinds of biased subset. A normal distribution which is shifted indicates some systematic bias which is independent of base position. Such a systematic bias creating a shifted normal distribution won’t be flagged as an error, since the tool cannot guess what the provided genome’s GC content should be.

Issues in the GC content distribution usually indicate a problem with the library. Sharp peaks on an otherwise smooth distribution are normally the result of a specific contaminant (adapter dimers for example), which may well be picked up by the over-represented sequences module. Broader peaks may represent contamination with a different species.

---

### Per-base N content

![A line graph of N content across all bases. It shows several peaks to 65 at specific positions and goes to 100 near the end.](../../images/per_base_N_content.png)

???
Sequences can contain the ambiguous base N for positions that could not be identified as a particular base. A high number of Ns can be a sign for a low quality sequence or even dataset. If no quality scores are available, the sequence quality can be inferred from the percent of Ns found in a sequence or dataset.

If a sequencer is unable to make a base call with sufficient confidence then it will normally substitute an N rather than a conventional base call. It’s not unusual to see a very low proportion of Ns appearing in a sequence, especially nearer the end of a sequence. However, if this proportion rises above a few percent it suggests that the analysis pipeline was unable to interpret the data well enough to make valid base calls.

---

### Sequence length distribution

![A graph with a single peak at 75, and 0 outside of this region.](../../images/sequence_length_distribution.png)

???
Some high throughput sequencers generate sequence fragments of uniform length, while others can output reads of wildly varying lengths. The length distribution can be then used as quality measure. You would expect a normal distribution for the best result. However, most sequencing results show a slowly increasing and then a steep falling distribution.

FastQC generates a graph showing the distribution of fragment sizes in the file which was analysed. In many cases this will produce a simple graph showing a peak only at one size, but for variable length FASTQ files this will show the relative amounts of each different size of sequence fragment.

This module will raise a warning if all sequences are not the same length. This module will raise an error if any of the sequences have zero length.

---

### Tag sequences: Adapter contamination

![The graph shows a line at zero for the five possible datasets.](../../images/adapter_content.png)

???
Tag sequences are artifacts at the ends of sequence reads such as multiplex identifiers, adapters, and primer sequences that were introduced during pre-amplification with primer-based methods. The base frequencies across the reads present an easy way to check for tag sequences. If the distribution seems uneven (high frequencies for certain bases over several positions), it could indicate some residual tag sequences. This doesn’t indicate a problem as such - just that the sequences will need to be adapter trimmed before proceeding with any downstream analysis.

To investigate tag or adapter content, FastQC generates a plot showing a cumulative percentage count of the proportion of the library which has seen each of the adapter sequences at each position. Once a sequence has been seen in a read it is counted as being present right through to the end of the read so the percentages you see will only increase as the read length goes on.

---


## Improving the quality of sequences

- Filtering of sequences
    - with small mean quality score
    - too small
    - with too many N bases
    - based on their GC content
    - ...
- Cutting/Trimming sequences
    - from low quality score parts
    - tails
    - ...
